lib/controller-inputs/FrameController.cpp

```diff
@@ -0,0 +1,223 @@
+#include "FrameController.h"
+
+using namespace FrameProtocol;
+
+FrameController::FrameController(RawSerial* dest) : m_dest(dest) {
+  for (unsigned i = 0; i < MAX_NUM_REDIRECT_FRAME; i++) {
+    m_sendSerialThreads[i] = new Thread(osPriorityNormal, SERIAL_THREAD_STACK_SIZE, m_sendSerialThreadMems[i]);
+  }
+}
+
+FrameController::~FrameController() {
+  for (unsigned i = 0; i < MAX_NUM_REDIRECT_FRAME; i++) {
+    delete m_sendSerialThreads[i];
+  }
+}
+
+void FrameController::sendFrame(DigitalFrameConfig& frameConfig) {
+  static uint8_t m_digitalFrame[FRAME_SIZE_DIGITAL] = {0};
+
+  m_digitalFrame[0]   = SOF_DIGITAL;
+  uint8_t second_byte = 0;
+
+  switch (frameConfig.boardType) {
+    case BoardType::MASTER:
+      second_byte |= MASTER_BOARD;
+      break;
+
+    case BoardType::ARM:
+      second_byte |= ARM_BOARD;
+      break;
+
+    case BoardType::SCIENCE:
+      second_byte |= SCIENCE_BOARD;
+      break;
+
+    default:
+      second_byte |= MASTER_BOARD;
+      break;
+  }
+
+  switch (frameConfig.inputType) {
+    case DigitalInputType::BTN:
+      second_byte |= BTN_INPUT;
+      break;
+
+    case DigitalInputType::SWITCH:
+      second_byte |= SWITCH_INPUT;
+      break;
+
+    default:
+      second_byte |= BTN_INPUT;
+      break;
+  }
+
+  m_digitalFrame[1] = second_byte;
+
+  uint16_t invalid_bits = 0, digital_values = 0;
+
+  invalid_bits = frameConfig.inputGroup.getValuesAndInvalidReads(digital_values);
+
+  *(uint16_t*)(&m_digitalFrame[2]) = digital_values;
+  *(uint16_t*)(&m_digitalFrame[4]) = invalid_bits;
+
+  // The serial port to send data out is a shared resource
+  m_mutex.lock();
+  for (unsigned int i = 0; i < FRAME_SIZE_DIGITAL; i++) m_dest->putc(m_digitalFrame[i]);
+  m_mutex.unlock();
+}
+
+void FrameController::sendFrame(AnalogFrameConfig& frameConfig) {
+  static uint8_t m_analogFrame[FRAME_SIZE_ANALOG] = {0};
+
+  static uint16_t analog_values[AnalogInputGroup::MAX_INPUTS_NUM] = {0};
+
+  m_analogFrame[0]    = SOF_ANALOG;
+  uint8_t second_byte = 0;
+
+  switch (frameConfig.boardType) {
+    case BoardType::MASTER:
+      second_byte |= MASTER_BOARD;
+      break;
+
+    case BoardType::ARM:
+      second_byte |= ARM_BOARD;
+      break;
+
+    case BoardType::SCIENCE:
+      second_byte |= SCIENCE_BOARD;
+      break;
+
+    default:
+      second_byte |= MASTER_BOARD;
+      break;
+  }
+
+  switch (frameConfig.inputType) {
+    case AnalogInputType::JOY:
+      second_byte |= JOY_INPUT;
+      break;
+
+    case AnalogInputType::POT:
+      second_byte |= POT_INPUT;
+      break;
+
+    default:
+      second_byte |= JOY_INPUT;
+      break;
+  }
+
+  m_analogFrame[1] = second_byte;
+
+  frameConfig.inputGroup.getValues(analog_values);
+
+  for (unsigned i = 0; i < frameConfig.inputGroup.getNumInputs(); i++) {
```

Comment:
fixed width int types for readability

Reply:
just wondering why is unsigned not readable? I find them expressive for the purpose of a for loop...maybe it's just a personal thing

Reply:
well my first reaction was unsigned what?  lol. its not immediately obvious that unsigned names a type (at least to me). more importantly its not something ive seen done often.